src/chalk-overview.md

```diff
@@ -0,0 +1,141 @@
+* [Negative reasoning in Chalk](http://aturon.github.io/blog/2017/04/24/negative-chalk/)
+* [Query structure in chalk](http://smallcultfollowing.com/babysteps/blog/2017/05/25/query-structure-in-chalk/)
+* [Cyclic queries in chalk](http://smallcultfollowing.com/babysteps/blog/2017/09/12/tabling-handling-cyclic-queries-in-chalk/)
+* [An on-demand SLG solver for chalk](http://smallcultfollowing.com/babysteps/blog/2018/01/31/an-on-demand-slg-solver-for-chalk/)
+
+## Parsing
+
+Chalk is designed to be incorporated with the Rust compiler, so the syntax and
+concepts it deals with heavily borrow from Rust. It is convenient for the sake
+of testing to be able to run chalk on its own, so chalk includes a parser for a
+Rust-like syntax.
+
+The parser takes that syntax and produces an [Abstract Syntax Tree (AST)][ast].
+You can find the [complete definition of the AST][chalk-ast] in the source code.
+
+The syntax contains things from Rust that we know and love for example traits,
+impls, and struct definitions. Parsing is often the first "phase" of
+transformation that a program goes through in order to become a format that
+chalk can understand.
+
+## Lowering
+
+After parsing, there is a "lowering" phase. This aims to convert traits/impls
+into "program clauses". A [`ProgramClause` (source code)][programclause] is
+essentially one of the following:
+
+* A [clause] of the form `consequence :- conditions` where `:-` is read as
+  "if" and `conditions = cond1 && cond2 && ...`
+* A universally quantified clause of the form
+  `forall<T> { consequence :- conditions }`
+  * `forall<T> { ... }` is used to represent [universal quantification]. See the
+    section on [Lowering to logic][lowering-forall] for more information.
+  * A key thing to note about `forall` is that we don't allow you to "quantify"
+    over traits, only types and regions (lifetimes). That is, you can't make a
+    rule like `forall<Trait> { u32: Trait }` which would say "`u32` implements
+    all traits". You can however say `forall<T> { T: Trait }` meaning "`Trait`
+    is implemented by all types".
+  * `forall<T> { ... }` is represented in the code using the [`Binders<T>`
+    struct][binders-struct].
+
+This is the phase where we encode the rules of the trait system into logic. For
+example, if we have:
+
+```rust,ignore
+impl<T: Clone> Clone for Vec<T> {}
+```
+
+We generate:
+
+```rust,ignore
+forall<T> { (Vec<T>: Clone) :- (T: Clone) }
+```
+
+This rule dictates that `Vec<T>: Clone` is only satisfied if `T: Clone` is also
+satisfied (i.e. "provable").
+
+### Well-formedness checks
+
+As part of lowering from the AST to the internal IR, we also do some "well
+formedness" checks. See the [source code][well-formedness-checks] for where
+those are done. The call to `record_specialization_priorities` checks
+"coherence" which means that it ensures that two impls of the same trait for the
+same type cannot exist.
```

Comment:
I'm guessing that this actually checks that specialization rules are not violated (i.e. overlapping impls are "valid")? In other words, if we wanted to implement specialization, would we just change the implementation of `record_specialization_priorities`?

Reply:
@nikomatsakis can answer this for sure, but my understanding is that this function is named this way because it will eventually help with specialization even though that isn't fully implemented yet. The description about the coherence checks is still correct.